pkg/registry/core/service/strategy.go

```diff
@@ -302,6 +303,43 @@ func (serviceStatusStrategy) WarningsOnUpdate(ctx context.Context, obj, old runt
 	return nil
 }
 
+// patchAllocatedValues allows clients to avoid a read-modify-write cycle while
+// preserving values that we allocated on their behalf.  For example, they
+// might create a Service without specifying the ClusterIP, in which case we
+// allocate one.  If they resubmit that same YAML, we want it to succeed.
+func patchAllocatedValues(newSvc, oldSvc *api.Service) {
+	if needsClusterIP(oldSvc) && needsClusterIP(newSvc) {
+		if newSvc.Spec.ClusterIP == "" {
+			newSvc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
+		}
+		if len(newSvc.Spec.ClusterIPs) == 0 {
+			newSvc.Spec.ClusterIPs = oldSvc.Spec.ClusterIPs
+		}
+	}
+
```

Comment:
it's an edge case, but shouldn't this avoid copying node ports that are explicitly referenced in the new service?

otherwise a user that depends on allocated ports initially then applies an update that explicitly uses a specific port that is not used by any other service can get an error when they would otherwise succeed:

1. create service with multiple ports, requiring a node port and healthcheck node port and depending on auto-allocated values

```yaml
apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  type: LoadBalancer
  externalTrafficPolicy: Local
  ports:
    - name: port1
      port: 100
    - name: port2
      port: 200
    - name: port3
      port: 300
  selector:
    test: "true"
```

resulting in 

```yaml
apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.0.0.178
  clusterIPs:
  - 10.0.0.178
  externalTrafficPolicy: Local
  healthCheckNodePort: 31091
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: port1
    nodePort: 32545
    port: 100
    protocol: TCP
    targetPort: 100
  - name: port2
    nodePort: 31089
    port: 200
    protocol: TCP
    targetPort: 200
  - name: port3
    nodePort: 32697
    port: 300
    protocol: TCP
    targetPort: 300
  selector:
    test: "true"
  sessionAffinity: None
  type: LoadBalancer
```

2. update service, explicitly specifying some ports and leaving others blank

```
apiVersion: v1
kind: Service
metadata:
  name: test
spec:
  type: LoadBalancer
  externalTrafficPolicy: Local
  ports:
    - name: port1
      port: 100
    - name: port2
      port: 200
      nodePort: 32545 # happens to match auto-allocated port1
    - name: port3
      port: 300
      nodePort: 31091 # happens to match auto-allocated health check port
  selector:
    test: "true"
```

result:

```
The Service "test" is invalid: spec.ports[1].nodePort: Duplicate value: 32545
```

The current code copies the previously allocated node port from the old service to the new service and errors because the same node port was used twice in the update

Reply:
but 32545 was already allocated, why we should deallocate it from port1 in your example?

Reply:
things the user sets explicitly should take priority over copying previously allocated values from the old service into the new to save reallocating

Reply:
so, empty doesn't mean "keep the old value", means "keep the old value if I'm not reusing it"?